Make STS rules above UTXO era-independent.

nc6 opened the pull request:
Honestly it's embarassing that I failed to notice that this worked. We
carry a few more constraints in the STS instances and Embed instances.
But the payoff is that we only have to override the specific rules we
want to change in each era, and not re-plumb everything.

Obviously this means that, in the future, if we make a change in a
different rule then we have to go back and turn the era-independent one
into an era-dependent version (and add instances for all intermediate
eras). But still, this feels like a definite improvement.

nc6 committed:
Make STS rules above UTXO era-independent.

Honestly it's embarassing that I failed to notice that this worked. We
carry a few more constraints in the STS instances and Embed instances.
But the payoff is that we only have to override the specific rules we
want to change in each era, and not re-plumb everything.

Obviously this means that, in the future, if we make a change in a
different rule then we have to go back and turn the era-independent one
into an era-dependent version (and add instances for all intermediate
eras). But still, this feels like a definite improvement.

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Mempool.hs:
```diff
@@ -108,9 +113,6 @@ applyTxs ::
   forall era m.
   ( STS (LEDGERS era),
     BaseM (LEDGERS era) ~ ShelleyBase,
```
```suggestion
```
We can remove the line above, it compiles for me without it.

Or am I missing something? :thinking:

nc6 replied:
Hmm, interesting. I wonder why I didn't see an "unneeded" warning.

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -127,9 +126,6 @@ applyBlockTransition ::
   forall era m.
   ( STS (STS.BBODY era),
     BaseM (STS.BBODY era) ~ ShelleyBase,
```
Same

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -164,10 +160,7 @@ applyBlockTransition globals state blk =
-    Environment (STS.BBODY era) ~ STS.BbodyEnv era,
-    State (STS.BBODY era) ~ STS.BbodyState era,
-    Signal (STS.BBODY era) ~ Block era
+    BaseM (STS.BBODY era) ~ ShelleyBase
```
Same

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxow.hs:
```diff
@@ -310,9 +321,28 @@ utxoWitnessed =
+    State (UTXOW era) ~ UTxOState era,
+    Signal (UTXOW era) ~ Tx era,
+    PredicateFailure (UTXOW era) ~ UtxowPredicateFailure era,
+    DSignable (Crypto era) (Hash (Crypto era) EraIndependentTxBody),
```
Duplicate of L328.

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxow.hs:
```diff
@@ -310,9 +321,28 @@ utxoWitnessed =
         TRC (UtxoEnv slot pp stakepools genDelegs, u, tx)
 
 instance
```
It seems that each constraint on the `STS` instance of `UTXOW` is needed here too. I haven't look at `Embed` closely, but it's kind of strange that we really need *all* of these constraints just to embed one error into another.

nc6 replied:
`Embed` has a superclass constraint of `STS`

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Bbody.hs:
```diff
@@ -183,9 +175,14 @@ bbodyTransition =
+    c ~ Crypto era,
+    DSignable c (Hash c EraIndependentTxBody),
+    ShelleyBased era,
+    DSignable c (Hash c EraIndependentTxBody),
```
Duplicates

mrBliss reviewed (commented):
Great! I'll try to update the consensus integration to see the full effect.

JaredCorduan reviewed (approved):
this is good news! thanks @nc6

mrBliss commented:
Huh, `Haskell CI / build (8.10.2, ubuntu-latest) (push)` failed because of a failure in the `cardano-prelude-test` testsuite. That testsuite should be disabled, IMO :slightly_smiling_face:

mrBliss commented:
bors merge

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxow.hs:
```diff
@@ -310,9 +316,12 @@ utxoWitnessed =
-    DSignable c (Hash c EraIndependentTxBody)
+  ( Era era,
+    ShelleyBased era,
+    STS (UTXOW era),
```
I think this constraint is causing CI to time out in https://hydra.iohk.io/build/4513046/nixlog/1.
The test case in question:
```haskell
testCase "don't spend from a bootstrap address" testBootstrapNotSpending
```

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Chain.hs:
```diff
@@ -372,46 +369,42 @@ chainTransition =
+    STS (BBODY era)
   ) =>
-  Embed (BBODY (ShelleyEra c)) (CHAIN (ShelleyEra c))
+  Embed (BBODY era) (CHAIN era)
```
Is this right?

Note that the `STS (CHAIN era)` instance above has a `Embed (BBODY era) (CHAIN era)` constraint, i.e., this instance. Since this instance has a `STS (CHAIN era)` constraint, there is a cycle.

Embed (BBODY era) (CHAIN era),

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Chain.hs:
```diff
@@ -372,46 +369,42 @@ chainTransition =
-    VRF.Signable (VRF c) Seed
+  ( Era era,
+    ShelleyBased era,
+    STS (CHAIN era)
```
Same?

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Chain.hs:
```diff
@@ -372,46 +369,42 @@ chainTransition =
-    VRF.Signable (VRF c) Seed
+  ( Era era,
+    ShelleyBased era,
+    STS (CHAIN era),
```
Same?

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Chain.hs:
```diff
@@ -372,46 +369,42 @@ chainTransition =
+  ( Era era,
+    c ~ Crypto era,
+    ShelleyBased era,
+    STS (CHAIN era),
```
Same?

mrBliss reviewed (commented):
I think there are still some cycles in the instances (`STS` vs. `Embed`), right?

mrBliss commented:
> I think there are still some cycles in the instances (`STS` vs. `Embed`), right?

Hmm, I'm not sure whether these cycles are problematic. I'm just suspicious because of the tests timing out.

mrBliss committed:
Remove cycles between Embed and STS instances

`Embed sub super` has the `STS super` super-class constraint, but `Embed sub
super` is almost always used as a super-class constraint of `Embed sub super`
itself. This is a cycle. My hypothesis is that these cycles are fine, *until*
you start relying on constraints that are implied by them, because the
corresponding dictionaries will be recursive too. I believe this is what caused
some tests to hang.

To remedy this, remove the `STS super` super-class constraint from the `Embed`
class. Next, remove the new redundant constraints and add the constraints that
are now missing.

Finally, remove the cycle between each pair of `Embed` and `STS` instances. For
example:

```diff
instance
  ( ..
    Embed (UTXO era) (UTXOW era),
    ..
  ) =>
  STS (UTXOW era)
  where
  ..

instance
  ( ..
-   STS (UTXOW era),
    ..
  ) =>
  Embed (UTXO era) (UTXOW era)
  where
  ..
```

mrBliss commented:
I just pushed a commit removing – what I think – are cycles between `Embed` and `STS` instances that could be causing the timeouts in the tests. It at least fixed the `don't spend from a bootstrap address` test locally. Let's see whether CI agrees.

mrBliss commented:
bors merge
